File: object.js
const state = require("./state");
const Node = require("./node");
const Identifier = require("./identifier");
const $EXP = require("./$expression");
const Instruction = require("./instruction");
const LET = require("./let");
const Scope = require("./scope");

module.exports = class OBJECT extends Node {
  constructor() {
    super();
    this.properties = {};
  }

  before() {
    if (this.name === undefined && this.object === undefined) {
      this.key = this.class.name.toLowerCase() + this.sequence;
      this.name = this.key;
    } else {
      this.key = Identifier.serialize(this);
    }
  }

  run(scope) {
    let name = this.key;

    state.assign(scope, name, `new state.${this.class.name}()`);
    scope.object = this;

    let list = [];

    for (let i = 0; i < this.class.args.length; i++) {
      let local = new LET();
      local.name = this.class.args[i];

      if (this.args[i] !== undefined) {
        let context = $EXP(this.args[i], 0);
        local.value = context.statement.run();
        list.push(local);
      } else {
        let context = $EXP("undefined", 0);
        local.value = context.statement.run();
        list.push(local);
      }
    }

    if (this.class.construct !== undefined) {
      let construct = this.class.construct;
      let instruction = new Instruction(scope, construct, false, true, false);
      list.push(instruction);
    }

    for (let node in this.class.declarations) {
      let declaration = this.class.declarations[node];
      let scope = new Scope();
      scope.instances[this.class.name] = this;
      list.push(new Instruction(scope, declaration, true, true, true, true));
    }

    if (this.object === undefined) {
      let context = $EXP(this.class.name + "s.push ( " + this.name + " )", 0);
      list.push(context.statement);

      state.run(scope, `state.${name}.id="${name}"`);

      context = $EXP(this.name, 0);
      list.push(context.statement);
    }

    return { value: name, next: list };
  }

  graph() {
    if (this.object !== undefined) this.object.properties[this.name] = this;
    this.class.instances[this.key] = this;
  }
};

File: python/src/nucleoid/nuc/OBJECT.py
"""
OBJECT runtime node class.
"""
from typing import Any, Dict, List
from .NODE import NODE


class OBJECT(NODE):
    """Object creation node."""

    def __init__(self, key: Any) -> None:
        """
        Initialize an OBJECT instance.

        Args:
            key: The object key
        """
        super().__init__(key)
        self.properties: Dict[str, Any] = {}
        self.name: Any = None
        self.object: Any = None
        self.class_ref: Any = None
        self.arguments: List[Any] = []

    def run(self, scope: Any) -> Dict[str, List[Any]]:
        """
        Execute object creation.

        Args:
            scope: The execution scope

        Returns:
            Dictionary with next instructions
        """
        from ..lang.ast import __Identifier__
        from ..lang.Evaluation import Evaluation
        from ..lang.estree.estree import append
        from ..state import state_instance
        from ..Instruction import Instruction
        from ..Scope import Scope
        from ...__nuc__ import __EXPRESSION__, __CALL__

        scope.object = self
        name = self.name

        # Determine variable path
        if self.object:
            variable = __Identifier__(
                append(self.object.resolve().node, self.name.node)
            )
        else:
            variable = self.name

        # Assign new instance
        state_instance.assign(
            scope,
            variable,
            Evaluation(f"new state.{self.class_ref.name}()"),
            False
        )

        # Assign ID
        state_instance.assign(
            scope,
            __Identifier__(f"{variable}.id"),
            Evaluation(f'"{name}"')
        )

        result_list: List[Any] = []

        # Handle non-nested objects
        if not self.object:
            state_instance.call(scope, f"{self.class_ref.list}.push", [f"state.{name}"])
            state_instance.assign(
                scope,
                __Identifier__(f'{self.class_ref.list}["{self.name}"]'),
                Evaluation(f"state.{name}")
            )

        # Call constructor if exists
        constructor = self.class_ref.methods.get("$constructor") if self.class_ref else None
        if constructor:
            call = __CALL__(
                constructor,
                [arg.node for arg in self.arguments]
            )
            result_list.append(call)

        # Process class declarations
        if self.class_ref:
            for node_key in self.class_ref.declarations:
                declaration = self.class_ref.declarations[node_key]
                decl_scope = Scope()
                decl_scope.dollar_instance = self

                result_list.append(
                    Instruction(
                        decl_scope,
                        declaration,
                        True,
                        True,
                        False,
                        False,
                        True,
                        True
                    )
                )
                result_list.append(
                    Instruction(
                        decl_scope,
                        declaration,
                        False,
                        False,
                        True,
                        True,
                        True,
                        True
                    )
                )

        # Add expression
        expression = Instruction(
            scope,
            __EXPRESSION__(variable.node),
            True,
            True,
            False,
            False
        )
        expression.derivative = False
        result_list.append(expression)

        return {"next": result_list}

    def graph(self) -> None:
        """Build dependency graph."""
        if self.object is not None:
            self.object.properties[str(self.name)] = self

        if self.class_ref:
            self.class_ref.instances[self.key] = self

    def resolve(self) -> Any:
        """
        Resolve object path.

        Returns:
            Identifier with full path
        """
        from ..lang.ast import __Identifier__
        from ..lang.estree.estree import append

        current = self
        resolved = self.name.node

        while current.object:
            current = current.object
            resolved = append(current.name.node, resolved)

        return __Identifier__(resolved)
